### PR TITLE
Support GHC JS backend

### PR DESCRIPTION
**Pull request by hamishmack**

### CODE REVIEW

**Comment by JoshMeredith** [comment]
Should we remove callbacks? That module has been upstreamed into `base`.

File: jsbits/jsstring.js [hunk 1]
```diff
@@ -853,37 +835,39 @@ function h$jsstringDecW64(hi,lo) {
...
 }
-#endif
 
 function h$jsstringHexI64(hi,lo) {
-    var lo0 = lo<0 ? lo+4294967296 : lo;
-    if(hi === 0) return lo0.toString(16);
-    return ((hi<0)?hi+4294967296:hi).toString(16) + h$jsstringHexIPadded8(lo0);
+    TRACE_JSSTRING("hexI64: " + hi + " " + lo);
+    var sign = (hi >>> 31) ? '-' : '';
+    var lo0 = sign ? ~(lo - 1) >>> 0 : lo;
```
Comment (doyougnu) [inline jsbits/jsstring.js:862]: please add a comment about `>>> 0`. Something like:

```suggestion
    // unsigned right bitshift by zero to convert to UInt32
    var lo0 = sign ? ~(lo - 1) >>> 0 : lo;
```

File: jsbits/text.js [hunk 2]
```diff
@@ -1,38 +1,23 @@
...
-	var cc = u1[i];
-	a[k++] = cc;
-	if(k === 60000) {
-	    s += String.fromCharCode.apply(this, a);
-	    k = 0;
-	    a = [];
-	}
-    }
-    return s + String.fromCharCode.apply(this, a);
+    return new TextDecoder().decode(arr.u8.subarray(off, off+len));
```
Comment (doyougnu) [inline jsbits/text.js:9]: *chef's kiss*

File: test/Tests/Properties.hs [hunk 3]
```diff
@@ -521,7 +521,7 @@ tests =
       testProperty "sf_cons" sf_cons,
       testProperty "j_cons" j_cons,
       testProperty "s_snoc" s_snoc,
-      testProperty "j_snoc" j_snoc,
+--      testProperty "j_snoc" j_snoc, -- property `.f`
```
Comment (doyougnu) [inline test/Tests/Properties.hs:524]: Why is this removed?

**Review by doyougnu — APPROVED** [review]
Well done Josh! I've tried to review best I could, but it could probably benefit from a more thorough review from someone who's worked on ghcjs. IMO it LGTM, just some minor fixes.

File: JavaScript/JSON/Types/Instances.hs [hunk 4]
```diff
@@ -1064,15 +1063,6 @@ realFloatToJSON d
     | otherwise               = doubleValue (realToFrac d)
 {-# INLINE realFloatToJSON #-}
 
-scientificToNumber :: Scientific -> Number
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply JavaScript/JSON/Types/Instances.hs:1067]: I don't remember this too well. I see that it was used in an instance that's commented out, and it's not exported - so I'd guess that I removed it because of an unused function warning.

File: JavaScript/JSON/Types/Internal.hs [hunk 5]
```diff
@@ -324,5 +323,5 @@ foreign import javascript unsafe
   js_listAssocs :: SomeObject m -> Exts.State# s -> (# Exts.State# s, Exts.Any {- [(JSString, Value)] -} #)
 
 foreign import javascript unsafe
-  "JSON.stringify($1)"
+  "JSON.stringify"
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply JavaScript/JSON/Types/Internal.hs:326]: Our current codegen means that the arguments are essentially appended on as a string, so that `JSON.stringify` becomes `JSON.stringify(args)` at use sites. It would be possible to use the fat arrow syntax, but redundant.

File: ghcjs-base.cabal [hunk 6]
```diff
@@ -1,6 +1,6 @@
 cabal-version:       3.0
 name:                ghcjs-base
-version:             0.2.1.0
+version:             1.0.0.0
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply ghcjs-base.cabal:3]: @hamishmack suggested this on slack a while ago as a way to break version bounds with projects using GHCJS. IMO, there's also the possibility of releasing with a new name, but I don't think it's necessary.

**Comment by luite** [comment]
I had a quick read already, but will do a proper review at the end of today

File: test/Tests/Properties.hs (hunk 3 above)
Comment (JoshMeredith) [reply test/Tests/Properties.hs]: I tested it locally and it works so I've re-enabled it.

File: JavaScript/Web/MessageEvent.hs [hunk 7]
```diff
@@ -38,6 +38,7 @@ getData me = case js_getData me of
 -- -----------------------------------------------------------------------------
 
 foreign import javascript unsafe
-  "$r2 = $1.data;\
-  \$r1 = typeof $r2 === 'string' ? 1 : ($r2 instanceof ArrayBuffer ? 3 : 2)"
+  "((x) => { var r2 = x.data;\
+           \ var r1 = typeof r === 'string' ? 1 : (r instanceof ArrayBuffer ? 3 : 2);\
```
[the remark this answers is not in the document]
Comment (JoshMeredith) [reply JavaScript/Web/MessageEvent.hs]: Thanks for the catch, I've pushed a fix.

File: ghcjs-base.cabal (hunk 6 above)
[the remark this answers is not in the document]
Comment (hamishmack) [reply ghcjs-base.cabal:3]: I think the problem with a new name is that they should be mutually exclusive.  If we keep the same name there is no way anyone will wind up with both in their build plan.

I wonder if we should use 0.8 or something to avoid confusion.  It seems unlikely we will want to release any more non `arch(javascript)` versions.
